### PR TITLE
tpl: Fix race condition in regexp cache

### LINKED ISSUE #1973: Data race in replaceRE
```
==================
WARNING: DATA RACE
Read by goroutine 51:
  regexp.(*Regexp).replaceAll()
      /usr/local/go/src/regexp/regexp.go:496 +0xad
  regexp.(*Regexp).ReplaceAllString()
      /usr/local/go/src/regexp/regexp.go:462 +0x18e
  github.com/spf13/hugo/tpl.replaceRE()
      /Users/bep/go/src/github.com/spf13/hugo/tpl/template_funcs.go:1275 +0x35f
  runtime.call128()
      /usr/local/go/src/runtime/asm_amd64.s:474 +0x50
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:303 +0xcd
  text/template.(*state).evalCall()
      /usr/local/go/src/text/template/exec.go:641 +0xd0b
  text/template.(*state).evalFunction()
      /usr/local/go/src/text/template/exec.go:515 +0x3ba
  text/template.(*state).evalCommand()
      /usr/local/go/src/text/template/exec.go:412 +0x2bc
  text/template.(*state).evalPipeline()
      /usr/local/go/src/text/template/exec.go:385 +0x1e8
  text/template.(*state).walk()
      /usr/local/go/src/text/template/exec.go:215 +0x1bc
  text/template.(*state).walk()
      /usr/local/go/src/text/template/exec.go:223 +0x9ed
  text/template.(*state).walkTemplate()
      /usr/local/go/src/text/template/exec.go:368 +0x57b
  text/template.(*state).walk()
      /usr/local/go/src/text/template/exec.go:228 +0x897
  text/template.(*state).walk()
      /usr/local/go/src/text/template/exec.go:223 +0x9ed
  text/template.(*Template).Execute()
      /usr/local/go/src/text/template/exec.go:178 +0x3c1
  html/template.(*Template).Execute()
      /usr/local/go/src/html/template/template.go:104 +0xc8
  github.com/spf13/hugo/hugolib.(*Site).renderThing()
      /Users/bep/go/src/github.com/spf13/hugo/hugolib/site.go:2065 +0xe8
  github.com/spf13/hugo/hugolib.(*Site).render()
      /Users/bep/go/src/github.com/spf13/hugo/hugolib/site.go:2039 +0x339
  github.com/spf13/hugo/hugolib.(*Site).renderAndWritePage()
      /Users/bep/go/src/github.com/spf13/hugo/hugolib/site.go:1966 +0x14b
  github.com/spf13/hugo/hugolib.pageRenderer()
      /Users/bep/go/src/github.com/spf13/hugo/hugolib/site.go:1442 +0x25a

Previous write by goroutine 52:
  regexp.compile()
      /usr/local/go/src/regexp/regexp.go:183 +0x1c1
  regexp.Compile()
      /usr/local/go/src/regexp/regexp.go:129 +0x50
  github.com/spf13/hugo/tpl.replaceRE()
      /Users/bep/go/src/github.com/spf13/hugo/tpl/template_funcs.go:1267 +0x1fa
  runtime.call128()
      /usr/local/go/src/runtime/asm_amd64.s:474 +0x50
  reflect.Value.Call()
      /usr/local/go/src/reflect/value.go:303 +0xcd
  text/template.(*state).evalCall()
      /usr/local/go/src/text/template/exec.go:641 +0xd0b
  text/template.(*state).evalFunction()
      /usr/local/go/src/text/template/exec.go:515 +0x3ba
  text/template.(*state).evalCommand()
      /usr/local/go/src/text/template/exec.go:412 +0x2bc
  text/template.(*state).evalPipeline()
      /usr/local/go/src/text/template/exec.go:385 +0x1e8
  text/template.(*state).walk()
      /usr/local/go/src/text/template/exec.go:215 +0x1bc
  text/template.(*state).walk()
      /usr/local/go/src/text/template/exec.go:223 +0x9ed
  text/template.(*state).walkTemplate()
      /usr/local/go/src/text/template/exec.go:368 +0x57b
  text/template.(*state).walk()
      /usr/local/go/src/text/template/exec.go:228 +0x897
  text/template.(*state).walk()
      /usr/local/go/src/text/template/exec.go:223 +0x9ed
  text/template.(*Template).Execute()
      /usr/local/go/src/text/template/exec.go:178 +0x3c1
  html/template.(*Template).Execute()
      /usr/local/go/src/html/template/template.go:104 +0xc8
  github.com/spf13/hugo/hugolib.(*Site).renderThing()
      /Users/bep/go/src/github.com/spf13/hugo/hugolib/site.go:2065 +0xe8
  github.com/spf13/hugo/hugolib.(*Site).render()
      /Users/bep/go/src/github.com/spf13/hugo/hugolib/site.go:2039 +0x339
  github.com/spf13/hugo/hugolib.(*Site).renderAndWritePage()
      /Users/bep/go/src/github.com/spf13/hugo/hugolib/site.go:1966 +0x14b
  github.com/spf13/hugo/hugolib.pageRenderer()
      /Users/bep/go/src/g
```

### PR DESCRIPTION
**Pull request by moorereason**
Protect regular expression cache with a mutex.

Fixes #1973

### CODE REVIEW

**Comment by moorereason** [comment]
wercker is dumb.  Failed on `go get` stage.

File: tpl/template_funcs.go [hunk 1]
```diff
@@ -1244,7 +1245,25 @@ func replace(a, b, c interface{}) (string, error) {
 	return strings.Replace(aStr, bStr, cStr, -1), nil
 }
 
-var regexpCache = make(map[string]*regexp.Regexp)
+// regexpCache represents a cache of regexp objects protected by a mutex.
+type regexpCache struct {
+	mu sync.Mutex
```
Comment (bep) [inline tpl/template_funcs.go]: Not that it matters too much in this case, but a RWMutex is better.

**Comment by moorereason** [comment]
@bep, I've reworked this PR, and it feels cleaner.  Hopefully, I got it right this time!  :smile:

**Comment by bep** [comment]
aedb13b

**Pull request closed without merging**